apps/src/dance/DanceAiClient.ts

```diff
@@ -0,0 +1,100 @@
+type EmojiAssociation = {
+  association: number[];
+};
+
+type CachedMapping = {
+  emojiAssociations: {[key: string]: EmojiAssociation};
+  output: string[];
+};
+
+const CachedPalettes: CachedMapping = UntypedCachedPalettes;
+const CachedBackgrounds: CachedMapping = UntypedCachedBackgrounds;
+const CachedForegrounds: CachedMapping = UntypedCachedForegrounds;
+
+// Output: JSON containing {backgroundColor:string, backgroundEffect:string, foregroundEffect:string}
+// ToDo: Break down into helper functions for each step rather than repeat code for each field
+export default function cachedAi(input: string) {
+  const inputWords: string[] = input.split(',');
+  const trimmedInput: string[] = inputWords.map(value => value.trim());
+  const inputPaletteVectors: {[key: string]: any} = {};
+  const inputBackgroundVectors: {[key: string]: any} = {};
+  const inputForegroundVectors: {[key: string]: any} = {};
+
+  const {emojiAssociations: paletteAssociations, output: palettes} =
+    CachedPalettes;
+  const {emojiAssociations: backgroundAssociations, output: backgrounds} =
+    CachedBackgrounds;
+  const {emojiAssociations: foregroundAssociations, output: foregrounds} =
+    CachedForegrounds;
+
+  // Pull individual association vectors for each input emoji
+  trimmedInput.forEach(value => {
+    if (Object.prototype.hasOwnProperty.call(paletteAssociations, value)) {
+      inputPaletteVectors[value] = paletteAssociations[value];
+    }
+    if (Object.prototype.hasOwnProperty.call(backgroundAssociations, value)) {
+      inputBackgroundVectors[value] = backgroundAssociations[value];
+    }
+    if (Object.prototype.hasOwnProperty.call(foregroundAssociations, value)) {
+      inputForegroundVectors[value] = foregroundAssociations[value];
+    }
+  });
+
+  // Sum word-association vectors of all input words
+  const sumPaletteAssociation: number[] = sumAssociationVectors(
+    inputPaletteVectors,
+    palettes.length
+  );
+  const sumBackgroundAssociation: number[] = sumAssociationVectors(
+    inputBackgroundVectors,
+    backgrounds.length
+  );
+  const sumForegroundAssociation: number[] = sumAssociationVectors(
+    inputForegroundVectors,
+    foregrounds.length
+  );
+
+  // Return max value of map vector as the output parameter e.g. backgroundColor="neon"
+  const finalColor: string =
+    palettes[
+      sumPaletteAssociation.indexOf(Math.max.apply(null, sumPaletteAssociation))
+    ];
+  const finalBackground: string =
+    backgrounds[
+      sumBackgroundAssociation.indexOf(
+        Math.max.apply(null, sumBackgroundAssociation)
+      )
+    ];
+  const finalForeground: string =
+    foregrounds[
+      sumForegroundAssociation.indexOf(
+        Math.max.apply(null, sumForegroundAssociation)
+      )
+    ];
+  const cachedAiResponse = {
+    backgroundEffect: finalBackground,
+    backgroundColor: finalColor,
+    foregroundEffect: finalForeground,
+  };
+
+  return JSON.stringify(cachedAiResponse);
```

Comment:
I was playing around with this and noticed that this function only returns the highest value of the second emoji. The first and third are ignored. Or, put a different way, it runs the correct calculations, but only as if the second emoji is all that was input by the user. I made some notes about this in a separate branch: https://github.com/code-dot-org/code-dot-org/pull/54217. I also took the opportunity to see if a different algorithm could produce correct results for the combination of all three emojis. The results of this investigation are also in the linked branch.

Reply:
I found out what was causing the error; there was a naive assumption that the input would strictly follow the format "[emoji1], [emoji2], [emoji3]" but the actual input was structured more like a naturalistic sentence. Therefore, when it would search the keys:
1 would not return due to having the phrase "Return a blah blah blah [emoji1]" and
3 would not return due to having an addition period at the end e.g. "silly." vs "silly"

Inputting JUST the emojiNames as a list in the expected format causes both algorithms to work.

In the interim, I'll merge your changes as described and see if we can incorporate some of the suggested JSON reformatting into the generation script itself rather than having the native file do this reformatting in 2.0

Reply:
(checking in from the airport)
Ahh that makes sense. I completely missed the `.` when I was looking at it. When I was thinking about it again this morning, I think I overdid the reformatting. I'd love to chat more on Monday about the best structure for the JSON file. 

I'm fine with merging as is now that we have an accurate algorithm. There's two changes I'd request however. First, I'd love to maintain both DanceAiClient and DanceAiClient2 so we can cross-reference the algorithms as we continue to refine the code here. However, they shouldn't both exist in production. So long as you don't delete your local branch (and/or I don't delete mine), they'll continue to exist in the git history. To merge this Pull Request, however, one should be removed in the PR (again, it'll be maintained in our local git histories, so we'll still be able to access it). As part of that, `DanceAiClient2` should be renamed to `DanceAiClient`. Additionally, there are a few comments that disable some code in `DanceAiModal` such as line 14 and line 139 that should be cleaned up.